Implement challenge Kubernetes resource deployment

detjensrobert opened the pull request:
The meat and potatoes, finally.

Depends on #38.

This deploys all the needed Kubernetes resources for challenges. Each challenge is created in its own namespace, with its `Deployment`s for all the pods defined in the challenge config, and any exposed services via `Service` LoadBalancers or `Ingress` configs.

Currently, non-HTTP challenges are exposed via their own separate LoadBalancer service at a DNS domain based on the challenge name. This approach does have a slightly higher cost, since this creates more cloud provider loadbalancers (on AWS each LB is about $20/month). We are also not currently exposing what domain to expose TCP challenges at, so this the name is generated from the challenge name.

This is slightly different than how the configuration was initially designed, where all TCP challenges would be exposed on the same domain at different ports. Doable, but using separate load balancers does not require messing with the ingress controller values to proxy TCP services. 

Switching from either of these to the Gateway API would let us keep a single loadbalancer (the Gateway controller) and direct TCP traffic via native ingress-like resources, at whatever domains we want, though each TCP service would still need to be on a different port.

detjensrobert committed:
Slugify challenge directory for container names and uploads

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Split kubernetes templates into own file, try rendering them

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Update challenge kube templates for actual templating

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Enable JSON features in minijinja

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Template out all challenge kube manifests, only print for now

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
move kubernetes deploy helpers to shared client mod

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Fix expose subdomain in example http challenge

This should be a subdomain, not the full domain.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Fix image pulls in test k3d cluster

Localhost in the k3d containers is not the host localhost, so it cant pull
images from the test registry. Instead, add the registry container to the
k3d network and alias the registry name to pull directly.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Fix ingress class name in ingress template

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Actually apply the challenge manifests to cluster

HTTP challenges should work now, but TCP challenges still need some work to
update the ingress config in order to get exposed.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Return created objects when applying kube manifest

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Skeleton for returning deploy expose results in struct

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Wait for deployed resources to become ready

This waits for:
- pods to become running
- deployments to complete rollouts
- ingresses to become published by controller
- LoadBalancer services to get external IP

Other types of resources or services are ignored and immediately return Ok.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Use inmemory no-op provider for external-dns for test deployment

We don't need to actually create records (nor is there a way to), but this will
still log the records it would have created.

Signed-off-by: Robert Detjens <github@detjens.dev>

detjensrobert committed:
Remove custom template env remnants

These were added to test overriding the minijinja `default` None handling.
Gating in the template was easier, and so this is not needed now.

Signed-off-by: Robert Detjens <github@detjens.dev>

KekoaM commented on src/clients.rs:
```diff
@@ -173,3 +179,174 @@ pub async fn kube_api_for(
+                            .ingress
+                            .as_ref()?
+                            .iter()
+                            // TODO: should this be any()? all controllers I've seen only add .ip here
```
I think this is fine as is here

KekoaM reviewed (approved):
All changes look good

detjensrobert committed:
Replace inline k8s condition funcs with upstreamed versions

These were contributed to the upstream kube library, so bump it and use
those contributed conditions instead of the custom inline ones.

Signed-off-by: Robert Detjens <github@detjens.dev>
